internal/gitmetadata.go

```diff
@@ -0,0 +1,113 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023 Datadog, Inc.
+
+package internal
+
+import (
+	"os"
+	"sync"
+)
+
+const (
+	// EnvGitMetadataEnabledFlag specifies the environment variable name for enable/disable
+	EnvGitMetadataEnabledFlag = "DD_TRACE_GIT_METADATA_ENABLED"
+	// EnvGitRepositoryURL specifies the environment variable name for git repository URL
+	EnvGitRepositoryURL = "DD_GIT_REPOSITORY_URL"
+	// EnvGitCommitSha specifies the environment variable name git commit sha
+	EnvGitCommitSha = "DD_GIT_COMMIT_SHA"
+	// EnvDDTags specifies the environment variable name global tags
+	EnvDDTags = "DD_TAGS"
+
+	// TagRepositoryURL specifies the tag name for git repository URL
+	TagRepositoryURL = "git.repository_url"
+	// TagCommitSha specifies the tag name for git commit sha
+	TagCommitSha = "git.commit.sha"
+
+	// TraceTagRepositoryURL specifies the trace tag name for git repository URL
+	TraceTagRepositoryURL = "_dd.git.repository_url"
+	// TraceTagCommitSha specifies the trace tag name for git commit sha
+	TraceTagCommitSha = "_dd.git.commit.sha"
+)
+
+var (
+	lock = &sync.Mutex{}
+
+	gitMetadataTags map[string]string
+)
+
+// Get git metadata from environment variables
+func getTagsFromEnv() map[string]string {
+	repositoryURL := os.Getenv(EnvGitRepositoryURL)
+	commitSha := os.Getenv(EnvGitCommitSha)
+
+	if repositoryURL == "" || commitSha == "" {
+		tags := ParseTagString(os.Getenv(EnvDDTags))
+
+		repositoryURL = tags[TagRepositoryURL]
+		commitSha = tags[TagCommitSha]
+	}
+
+	if repositoryURL == "" || commitSha == "" {
+		return nil
+	}
+
+	return map[string]string{
+		TagRepositoryURL: repositoryURL,
+		TagCommitSha:     commitSha,
+	}
+}
+
+// GetGitMetadataTags returns git metadata tags
+func GetGitMetadataTags() map[string]string {
+	if gitMetadataTags != nil {
+		return gitMetadataTags
+	}
+	lock.Lock()
+	defer lock.Unlock()
+	if gitMetadataTags != nil {
+		return gitMetadataTags
+	}
```

Comment:
⚠️ Is lock supposed to protect `gitMetadataTags`? If yes, why is it being accessed before the lock is taken here? `ResetGitMetadataTags` also fails to take the lock. Either this lock isn't needed, or it should be applied consistently.

Reply:
I used standard go singleton pattern there.
[like](https://refactoring.guru/design-patterns/singleton/go/example) [or]( https://thedevelopercafe.com/articles/singleton-in-golang-839d8610958b)
Is it wrong?

Reply:
Ah, good catch Felix. I didn't look at this part closely. There is a data race right now, since `ResetGitMetadataTags` can change `gitMetadataTags` while `GetGitMetadataTags` is also reading it without a lock. Example test to demonstrate the race below. The test will fail under the race detector even if you add a lock to `ResetGitMetadataTags`, and will also fail if you have two goroutines making calls to `GetGitMetadataTags` concurrently, so Go's race detector isn't convinced that this pattern is safe.

@sashacmc I'm assuming the intent here is to make the check cheap if the tags are already set, since this function will be called during span creation? If so, I think this is an un-needed optimization right now because the tracer function where this will be called is [already under a lock](https://github.com/DataDog/dd-trace-go/blob/ba17616e0f7abbe5d00379efad74f83e0fb6052c/ddtrace/tracer/spancontext.go#L299-L301). So it's not adding any additional contention to existing code to do the whole thing under the lock.

More to the point, concurrency is tricky and it's probably better as a rule of thumb to protect data that's intended to be accessed concurrently consistently across the board, and only resort to more clever patterns if it's absolutely needed _and_ you can be absolutely certain they're correct. :)

```go
// run this with `go test -race -run=TagsRace` to see the data race
func TestTagsRace(t *testing.T) {
	var wg sync.WaitGroup
	wg.Add(2)
	go func() {
		defer wg.Done()
		for i := 0; i < 1000; i++ {
			_ = GetGitMetadataTags()
		}
	}()
	go func() {
		defer wg.Done()
		for i := 0; i < 1000; i++ {
			// NB: you can make this also do "_ = GetGitMetadataTags()" and still get a race
			ResetGitMetadataTags()
		}
	}()
	wg.Wait()
}
```

Reply:
@nsrip-dd thanks for the explanation!
(currently ResetGitMetadataTags used only in test, but will better protect it for possible future improper usage)
Fixed!